R/convert_lines_pts.R

```diff
@@ -0,0 +1,133 @@
+#' }
+convert_lines_pts <- function(processed_lines = fs::path(PEMprepr::read_fid()$dir_20105020_clean_field_data$path_rel, "proc_s1_transects5.gpkg"),
+                              trast,
+                              buffer = 2.5,
+                              neighbours = FALSE,
+                              write_output = TRUE,
+                              out_dir = PEMprepr::read_fid()$dir_20105020_clean_field_data$path_rel,
+                              out_name = "allpoints.gpkg") {
+  # check processed_transects is path or spatvect
+  processed_lines <- PEMprepr:::read_sf_if_necessary(processed_lines)
+
+  # check trast is path or spatrast
+  trast <- PEMprepr:::read_spatrast_if_necessary(trast)
+  tname <- names(trast)
+
+  # check if ID column exists
+  if ("ID" %in% colnames(processed_lines) == FALSE) {
+    processed_lines <- processed_lines |>
+      dplyr::mutate(ID = seq(1, length(processed_lines$order), 1))
+  }
+
+  processed_transects_id <- sf::st_drop_geometry(processed_lines)
+
+  geom_type <- as.character(unique(sf::st_geometry_type(processed_lines, by_geometry = TRUE)))
+
+  if (geom_type %in% c("LINESTRING")) {
+    # for lines
+    lBuff <- sf::st_buffer(processed_lines, dist = buffer, endCapStyle = "FLAT", joinStyle = "MITRE")
+    lBuff <- sf::st_cast(lBuff, "MULTIPOLYGON")
+  } else {
+    # for points
+    lBuff <- sf::st_buffer(processed_lines, dist = buffer)
+    lBuff <- sf::st_cast(lBuff, "MULTIPOLYGON")
+  }
+
+  # extract the XY values of raster cells where crossed lines
+  xys <- terra::extract(trast, terra::vect(lBuff), xy = TRUE)
+
+  raster_points_xy <- xys |>
+    sf::st_as_sf(coords = c("x", "y"), crs = 3005) |>
+    terra::merge(processed_transects_id) |>
+    dplyr::select(c(-"ID", -tname))
+
+  # add slice and tid (transect id)
+  allpts <- raster_points_xy |>
+    dplyr::mutate(tid = tolower(gsub("_[[:alpha:]].*", "", .data$transect_id))) |>
+    dplyr::mutate(slice = sub(".*(?=.$)", "", gsub("\\..*", "", .data$tid), perl = T))
+
+  # add neighbours if selected
+
+  if (neighbours) {
+    sf::st_geometry(allpts) <- "geom"
+    cli::cat_line()
+    cli::cli_alert_warning("generating neighbouring points")
+
+    dat_pts <- allpts |> dplyr::mutate(ptsID = dplyr::row_number())
+
+    dat_atts <- sf::st_drop_geometry(dat_pts)
+
+    pts <- terra::vect(dat_pts)
+    cellNums <- terra::cells(trast, pts)
+    cell_lookup <- tibble::tibble(ID = pts$ptsID, cell = cellNums)
+
+    adjCells <- terra::adjacent(trast, cells = cellNums[, 2], directions = "queen", include = TRUE) |>
+      tibble::as_tibble(.name_repair = "unique") |>
+      dplyr::rename_with(~ c("Orig", paste("Adj", 1:8, sep = ""))) |>
+      dplyr::mutate(ID = dplyr::row_number())
+
+    adjLong <- adjCells |>
+      tidyr::pivot_longer(cols = !("ID"), names_to = "position", values_to = "CellNum") |>
+      dplyr::arrange("ID", "position")
+
+    terra::values(trast) <- 1:terra::ncell(trast)
+    cellnums <- 1:terra::ncell(trast)
```

Comment:
```suggestion
    terra::values(trast) <- cellnums <- seq_len(terra::ncell(trast))
```

Can simplify by assigning twice, and make safer by using `seq_len()`